src/Server/Model/AbstractToken.php

```diff
@@ -57,87 +58,119 @@
     protected $scopes = [];
 
     /**
-     * Set the token (either access or refresh token)
-     *
-     * @param  string $token
-     * @return void
+     * AbstractToken constructor.
      */
-    public function setToken($token)
+    private function __construct()
     {
-        $this->token = (string) $token;
     }
 
     /**
-     * Get the token
-     *
-     * @return string
+     * @param int                      $ttl
+     * @param TokenOwnerInterface|null $owner
+     * @param Client|null              $client
+     * @param null                     $scopes
+     * @return AbstractToken
      */
-    public function getToken(): string
-    {
-        return $this->token;
+    public static function generateNew(
+        int $ttl = 0,
+        TokenOwnerInterface $owner = null,
+        Client $client = null,
+        $scopes = null
+    ) {
+
+        if (is_array($scopes)) {
+            foreach ($scopes as &$scope) {
+                $scope = $scope instanceof Scope ? $scope->getName() : (string) $scope;
+            }
+        }
+
+        if (is_string($scopes)) {
+            $scopes = explode(' ', $scopes);
+        }
+
+        $class = get_called_class();
+        $token = new $class();
+
+        $token->token     = bin2hex(random_bytes(20));
+        $token->owner     = $owner;
+        $token->client    = $client;
+        $token->scopes    = $scopes ?? [];
+        $token->expiresAt = $ttl ? (new DateTimeImmutable())->modify("+$ttl seconds") : null;
+
+        return $token;
     }
 
     /**
-     * Set the client that issued this token
-     *
-     * @param  Client $client
-     * @return void
+     * @param array $data
+     * @return AbstractToken
      */
-    public function setClient(Client $client)
+    public static function reconstitute(array $data)
     {
-        $this->client = $client;
+        $data['owner']     = $data['owner'] ?? null;
```

Comment:
$data['scopes'] might not be set. but i should move it to the object props as you say.